### ISSUE TITLE
TST: Remaining GitHub Actions actions

### ISSUE THREAD

**Issue by pllim** [astropy/astropy#11038]
### Description

Remaining things to port from `.travis.yml`:

- [x] ~Caching~ with `ccache` and `.hypothesis`. (`.hypothesis` caching PR opened at #11161, @saimn is interested to look at the rest when he gets the chance). **UPDATE: Haven't seen a real need for caching so far.**
  - Formal documentation at https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows (there is a 5 GB hard limit)
  - There is a [blog about using ccache in Actions](https://cristianadam.eu/20200113/speeding-up-c-plus-plus-github-actions-using-ccache/) that maybe we can try.
  - ~Do we want to start caching pip dependencies too?~ See https://github.com/actions/cache/blob/main/examples.md#python---pip (The answer is "no" according to https://github.com/astropy/astropy/issues/11038#issuecomment-786770402 .)
- [x] `pytest` conditions in test modules using `TRAVIS` env var. See #11128
  - Example: https://github.com/astropy/astropy/blob/f9ba0f8e68abde11e5af05979213e3909c3f6fc7/conftest.py#L29
- [x] ~Slack notification~ (both on success and failure) for cron job. (@saimn tried something but it didn't work, @Cadair has a cunning plan at https://github.com/astropy/astropy/issues/11038#issuecomment-743229649.) **UPDATE: Nothing is happening on this front and no one has asked for it lately.**
- Cron:
  - [x] Big-endian s390x architecture with apt (`arch: s390x`, `ASTROPY_USE_SYSTEM_ALL=1`) (@astrofrog opened #11045 )
  - [x] ARM64 architecture with apt (`arch: arm64`, `ASTROPY_USE_SYSTEM_ALL=1`) (@astrofrog opened #11045 )
- [x] Finally, delete `.travis.yml`. (@astrofrog opened #11045 )

From `.circleci/config.yml`:

- [x] 32-bit and parallel (custom `astropy.cfg`, Python 3.7 `-n=4 --durations=50`, Python 3.8 `py38-test-double`). See #11043
- `image-tests-mpl302` (@Cadair says no need to port)
- `image-tests-mpl310` (@Cadair says no need to port)
- `image-tests-mpldev` (@Cadair says no need to port)

Things already in Actions we need to fix:

- [x] Allow remote data job to fail. See #11036
- [x] Separate cron jobs into daily and weekly. See #11036
- [x] See if we can add `--open-files` back to a newer job in future PR. See #11037
- [x] `codecov` uploads report fine but not reporting back to PR. How to fix? ~Try codecov/codecov-action#84 and see?~ Looks like a [a bug on their side](https://community.codecov.io/t/codecov-commit-status-check-missing-for-pr-when-using-codecov-action/1813) that is now fixed.
- [x] Do we still need branch filters? I guess there is no need for now but we can revisit if we change our minds.
- [x] Fix numpy-dev/remote data job failures. ~See #11042~ (Went away on its own when numpy-dev wheel updated.)
- [x] Fix cron job failures. (Might be the same problem as numpy-dev/remote data job above, ~so should wait until #11042 is merged first~ yes it was.)
- [x] Do we need to roll our own `[ci skip]`? Failed attempts at #11044 and #11129, maybe https://github.com/astropy/astropy/issues/11038#issuecomment-739075305 . See #11168. And now [GitHub supports it](https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/), so also see #11316 
- [x] Find a way to auto-cancel Actions, see https://github.com/astropy/astropy/issues/11038#issuecomment-727070545 . See #11168

Additional ideas:

- [x] Can we tell `ci_workflows.yml` to run Actions on the PR fork instead of here, but report back the status to the PR anyway? The astropy version reported will be wrong due to lack of tags on the fork and if some jobs require special tokens, they will not work. (@saimn will investigate and he actually meant **the opposite**, see #11054)
  - [x] ~Blocked by astropy/pytest-astropy-header#31~

### Additional context

I milestoned this issue because ideally we should sort everything out by 4.3 release.

See #10388 for more info (e.g., pros, cons, related links).

**Comment by Cadair** [comment]
I don't think the current implementation of artifacts on GH actions are up to replacing circle at this point.

**Comment by astrofrog** [comment]
We could probably move the 32-bit job over though?

**Comment by pllim** [comment]
Just to clarify, not everything in CircleCI has to port over. I am just putting them up there for discussion. "Won't fix" is an acceptable answer if current solution is sustainable and is working.

**Comment by saimn** [comment]
About the "cancel previous builds" feature, there are some actions to do that:
- https://github.com/styfle/cancel-workflow-action
- https://github.com/fkirc/skip-duplicate-actions (could also be used to skip some builds if only some paths were modified?)
- and others : https://github.com/marketplace?type=actions&query=cancel+workflow

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/issues/11038#issuecomment-727070545

Thanks, @saimn ! Personally, I still find it ridiculous that Actions does not support that natively and we have to rely on third-party apps.

**Comment by pllim** [comment]
@Zac-HD , any advise on how to cache `.hypothesis` properly on GitHub Actions? Thanks!

**Comment by astrofrog** [comment]
Can we use https://github.com/actions/cache for .hypothesis?

**Comment by Zac-HD** [comment]
Yep, that's how I'd do it!  

The simple option: just cache `.hypothesis`.  That's it; it'll work just fine.

More complex: you could also use the `restore-keys` option to ensure that the cache was specific to the current branch and/or Python version and/or Hypothesis version.  This *could* in principle help because it avoids Hypothesis evicting failing examples which are specific to a different job, but I'm not sure it's worth the extra complexity.

**Comment by embray** [comment]
Something I discovered since the switchover which is frustrating (if I understand correctly) it seems that each "check" performed by tox is performed in a single job.  So if one check fails (e.g. remote data) I can't re-run that one check since it's not a separate job.

*Update:* @pllim I see you already addressed this by adding a separate allowed_fail job containing, for now, the remote data tests.  That might be good enough for most cases.

I wonder if GH Actions allows reuse though, as that would seem to make it easier to create separate jobs with less duplication.  GitLab CI's file format is very nice in this regard; I have to study how reuse is enabled in GH Actions, if at all...

**Comment by embray** [comment]
OK, apparently YAML anchors are explicitly not supported by Actions, and this is still a [requested feature](https://github.community/t/support-for-yaml-anchors/16128/43).  I wonder why not.  It's a basic feature of YAML that YAML parsers for all major languages support.  What a let-down.  Use of anchors can be really nice in GitLab CI configs.

**Comment by astrofrog** [comment]
One can create composite actions (basically a set of steps to re-use) but frustratingly for now composite actions can't use actions, they can only include run steps (so no e.g. set up Python). Once that is fixed, we could use composite actions to reduce duplication.

**Comment by embray** [comment]
All the more reason it would be nice to have a simple feature of YAML that would solve the problem without any convolutions like "composite actions" :sweat_smile:

**Comment by mhvk** [comment]
Thanks all for the amazing work! My really wished for missing feature is `[skip ci]`...

**Comment by pllim** [comment]
Codecov action works just fine in spacetelescope/stsci-package-template#25 . Either they just fixed it or the bug only triggers under some condition. Should keep an eye out to see if it automagically works again or not.

**Comment by pllim** [comment]
Back to https://github.com/astropy/astropy/issues/11038#issuecomment-727070545 on using third-party apps to auto-cancel duplicate jobs (and in general), I am still hesitant because there is no way to vet the quality of the Actions code or whether it is malicious or not. Does benefit outweigh the risk? Still not supported natively... https://github.community/t/github-actions-cancel-redundant-builds-not-solved/16025/34

**Comment by pllim** [comment]
Re: skipping CI (not to be confused with cancelling duplicate runs) -- I implemented [action-skip-ci](https://github.com/pllim/action-skip-ci) with proof-of-concept at kali-maa/astropy#41 . Not pretty but it works. Maybe easier to customize than some third-party app written in JS?

**Comment by saimn** [comment]
I think it's fine to use a third-party app for a feature like skip-ci. For example https://github.com/styfle/cancel-workflow-action seems widely used and has ~80 lines of code, I don't think there is a huge risk.

**Comment by pllim** [comment]
On the topic of Slack notification, GitHub says they don't have plan to add official support in the near future (see integrations/slack#940). @Cadair suggested pushing to Matrix and have the bridge push to Slack. He has the Matrix portion working already for Sunpy, as follows, and willing to help us set up a bot account for this purpose:

https://github.com/sunpy/sunpy/blob/master/tools/send_matrix_message.xsh

https://github.com/sunpy/sunpy/blob/94b31a412fe9486db1a448e4e2e99a77d57ec2ae/azure-pipelines.yml#L162

**Comment by pllim** [comment]
Update: Just caching and Slack notifications left.

**Comment by jdavies-st** [comment]
>   * Do we want to start caching pip dependencies too? See https://github.com/actions/cache/blob/main/examples.md#python---pip

I will comment here that caching `pip` dependencies can mask things like build issues caused by newly-released build dependencies.  We saw this over at `jwst` where we were caching build dependencies, and it caused us to not catch when Numpy released 1.20, and caused users of 1.19x to have errors about API incompatibility between the built and runtime versions of Numpy.  Pip just uses the latest version of a PEP518 dependency like `numpy` or the latest cached version if it is not specifically limited in pyproject.toml.  See discussion here

https://github.com/spacetelescope/jwst/issues/5744#issuecomment-779405057

We did not catch this in our CI because we were caching wheels in the pip cache.  It showed up when run with `pip install --no-cache-dir` and also would have if we had turned off pip caching in our actions.

**Comment by saimn** [comment]
I don't understand how it is a cache issue. Read the issue you linked, it seems you were building wheels with the latest numpy (using `numpy` without pinning in `pyproject.toml`) which then indeed leads to wheels that are not compatible with older numpy. `oldest-supported-numpy` is the solution to this, or you could pin manually the numpy version used to build the wheels  in `pyproject.toml`. Typically you want to build wheels with the oldest numpy supported for each Python version, which `oldest-supported-numpy` does.

**Comment by jdavies-st** [comment]
With the cache, for some reason, `pip` builds the `drizzle` wheels with an older, cached version of `numpy`.  I'm not sure why.

Agree that `oldest-supported-numpy` is the current solution given the lack of tooling to deal with this in `pip`.

**Comment by embray** [comment]
For numpy in particular shouldn't there be a test job for testing against "latest numpy".  I agree with @saimn though that this is not an issue with caching.

**Comment by pllim** [comment]
Re: `jwst` -- Out of topic now, but testing a pipeline against latest numpy is a little excessive, no?
